File: src/main/java/org/cubectl/identity/auth/AuthService.java
package org.cubectl.identity.auth;

import java.util.Set;
import java.util.stream.Collectors;

import org.cubectl.identity.role.PermissionEntity;
import org.cubectl.identity.role.RoleEntity;
import org.cubectl.identity.role.RoleRepository;
import org.cubectl.identity.user.UserEntity;
import org.cubectl.identity.user.UserRepository;
import org.cubectl.identity.user.UserStatus;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
public class AuthService {

    private final UserRepository userRepository;
    private final RoleRepository roleRepository;
    private final PasswordEncoder passwordEncoder;
    private final JwtService jwtService;
    private final RefreshTokenService refreshTokenService;

    public AuthService(
            UserRepository userRepository,
            RoleRepository roleRepository,
            PasswordEncoder passwordEncoder,
            JwtService jwtService,
            RefreshTokenService refreshTokenService
    ) {
        this.userRepository = userRepository;
        this.roleRepository = roleRepository;
        this.passwordEncoder = passwordEncoder;
        this.jwtService = jwtService;
        this.refreshTokenService = refreshTokenService;
    }

    @Transactional
    public TokenResponse register(String username, String password) {
        if (userRepository.existsByUsername(username)) {
            throw new IllegalArgumentException("Username already exists");
        }
        boolean firstUser = userRepository.count() == 0;
        UserEntity user = userRepository.save(new UserEntity(username, passwordEncoder.encode(password)));
        roleRepository.findByCode(firstUser ? "admin" : "viewer").ifPresent(role -> user.getRoles().add(role));
        return issueTokens(user);
    }

    @Transactional
    public TokenResponse login(String username, String password) {
        UserEntity user = userRepository.findWithRolesByUsername(username)
                .orElseThrow(() -> new IllegalArgumentException("Invalid credentials"));
        if (user.getStatus() != UserStatus.ACTIVE || !passwordEncoder.matches(password, user.getPasswordHash())) {
            throw new IllegalArgumentException("Invalid credentials");
        }
        user.markLogin();
        return issueTokens(user);
    }

    @Transactional
    public TokenResponse refresh(String refreshToken) {
        UserEntity user = refreshTokenService.consume(refreshToken);
        return issueTokens(userRepository.findWithRolesById(user.getId()).orElseThrow());
    }

    @Transactional
    public void logout(String refreshToken) {
        if (refreshToken != null && !refreshToken.isBlank()) {
            refreshTokenService.revoke(refreshToken);
        }
    }

    private TokenResponse issueTokens(UserEntity user) {
        Set<String> permissions = permissions(user);
        return new TokenResponse(
                jwtService.issueAccessToken(user.getId(), user.getUsername(), permissions),
                refreshTokenService.create(user),
                "Bearer"
        );
    }

    private Set<String> permissions(UserEntity user) {
        return user.getRoles().stream()
                .flatMap(role -> role.getPermissions().stream())
                .map(PermissionEntity::getCode)
                .collect(Collectors.toUnmodifiableSet());
    }

    public record TokenResponse(String accessToken, String refreshToken, String tokenType) {
    }
}

File: src/main/kotlin/org/cubectl/identity/auth/AuthService.kt
package org.cubectl.identity.auth

import org.cubectl.identity.auth.dto.TokenResponse
import org.cubectl.identity.role.PermissionEntity
import org.cubectl.identity.role.RoleRepository
import org.cubectl.identity.user.UserEntity
import org.cubectl.identity.user.UserRepository
import org.cubectl.identity.user.UserStatus
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class AuthService(
    private val userRepository: UserRepository,
    private val roleRepository: RoleRepository,
    private val passwordEncoder: PasswordEncoder,
    private val jwtService: JwtService,
    private val refreshTokenService: RefreshTokenService,
) {

    @Transactional
    fun register(username: String, password: String): TokenResponse {
        if (userRepository.existsByUsername(username)) {
            throw IllegalArgumentException("Username already exists")
        }
        val firstUser = userRepository.count() == 0L
        val passwordHash = requireNotNull(passwordEncoder.encode(password)) { "Password encoder returned null" }
        val user = userRepository.save(UserEntity(username, passwordHash))
        roleRepository.findByCode(if (firstUser) "admin" else "viewer").ifPresent { role ->
            user.roles.add(role)
        }
        return issueTokens(user)
    }

    @Transactional
    fun login(username: String, password: String): TokenResponse {
        val user = userRepository.findWithRolesByUsername(username)
            .orElseThrow { IllegalArgumentException("Invalid credentials") }
        if (user.status != UserStatus.ACTIVE || !passwordEncoder.matches(password, user.passwordHash)) {
            throw IllegalArgumentException("Invalid credentials")
        }
        user.markLogin()
        return issueTokens(user)
    }

    @Transactional
    fun refresh(refreshToken: String): TokenResponse {
        val user = refreshTokenService.consume(refreshToken)
        return issueTokens(userRepository.findWithRolesById(user.id).orElseThrow())
    }

    @Transactional
    fun logout(refreshToken: String?) {
        if (!refreshToken.isNullOrBlank()) {
            refreshTokenService.revoke(refreshToken)
        }
    }

    private fun issueTokens(user: UserEntity): TokenResponse =
        TokenResponse(
            accessToken = jwtService.issueAccessToken(user.id, user.username, permissions(user)),
            refreshToken = refreshTokenService.create(user),
            tokenType = "Bearer",
        )

    private fun permissions(user: UserEntity): Set<String> =
        user.roles
            .flatMap { role -> role.permissions }
            .map(PermissionEntity::code)
            .toSet()
}
